### PR TITLE
Add field to rd_kafka_queue which denotes if it contains fetched msgs

### PR DESCRIPTION
**Pull request by milindl**
An issue in v2.1.0 was fixed in which max.poll.interval.ms was not honored,
    because it was reset on any queue poll, not just consumer poll.
    It was changed it so that only certain rdkafka.h functions which were polling
    would reset the timer.
    
   However, librdkafka exposes a method rd_kafka_queue_get_consumer, which returns
    the consumer queue, and the application can poll this queue for events rather
    than calling consume poll. There is no way to distinguish polls to this queue
    and an arbitrary queue, and it won't reset the timer.
    
   So, a new field is maintained inside the queue denoting if it might
    contain fetched messages, or not. It deals with forwarding of queues, so
    if a queue which receives fetched messages is forwarded multiple times,
    calling poll on the forwardee will also reset the timer.

Note that there are a few points where someone might forward/unforward into a queue _while_ polling it. (See cases with `does_q_contain_consumer_msgs = rd_kafka_q_consumer_cnt(rkq, 1);`) which might make the timer reset wrong on these occasions. 

I don't think it matters much since that's a somewhat convoluted use case and to fix it will require taking a lock over the duration of the entire poll. But let me know what you think.

Fixes [confluentinc/confluent-kafka-go#980](https://github.com/confluentinc/confluent-kafka-go/issues/980)

### CODE REVIEW

File: src/rdkafka_queue.c
Comment (pranavrth) [inline src/rdkafka_queue.c:132]: improve doc

File: src/rdkafka_queue.c
Comment (pranavrth) [inline src/rdkafka_queue.c:134]: remove this

File: src/rdkafka_queue.c
Comment (pranavrth) [inline src/rdkafka_queue.c:140]: Use warn log instead of assert

**Review by pranavrth — CHANGES_REQUESTED** [review]

**Review by pranavrth — COMMENTED** [review]
As we are locking on polling, please check performance once.

File: src/rdkafka_queue.c
Comment (pranavrth) [inline src/rdkafka_queue.c:151]: Shouldn't we release this lock earlier than above call? I think circular dependency case will cause a deadlock which will be very hard to identify even if it is a very rare scenario.

**Review by pranavrth — CHANGES_REQUESTED** [review]

File: src/rdkafka_queue.c
Comment (milindl) [reply src/rdkafka_queue.c:155]: That is not possible to do, I think
Consider the case that A, B are queues.. Someone calls forward(A,B). propagate(A) is called. After processing, suppose we unlock the mutex and then call propagate(B). 
Before propagate(B) starts to run, someone does forward(A, NULL) on another thread. If A->rkq_fwdq is the only reference to B, then B will get destroyed now, and we might encounter a segmentation fault on trying to access fields of B when propagate(B) finally runs.

Both these cases (circular dependency and concurrent calling of forward(A,B) and forward(A,NULL)) are rare, but while causing a circular dependency is an error case which the user should take care to avoid, concurrent forwarding is legitimate use case if someone is trying something with dynamic queue forwarding. So better to support the legitimate use case since we have to choose between them.

Changed the documentation for this.

File: src/rdkafka_queue.c
Comment (milindl) [reply src/rdkafka_queue.c]: Resolving this as we are moving to flag and never unsetting it.

File: src/rdkafka_queue.c
Comment (milindl) [reply src/rdkafka_queue.c]: Changed name to rd_kafka_q_consumer_propagate , since we only set the flag and there is no count anymore

File: src/rdkafka_queue.c
Comment (milindl) [reply src/rdkafka_queue.c]: Remove this statement entirely because we're now using the flag

File: src/rdkafka.h
Comment (emasab) [inline src/rdkafka.h:3430]: Here "loose" was correct, to mean "release"

File: src/rdkafka_queue.h
Comment (pranavrth) [inline src/rdkafka_queue.h:1186]: We can live without the lock if the `rkq` is a consumer queue because if the flag is already set, it is never unset. So if we first check the flag and it is already set, we don't need to lock the queue to get the value.

This benefits the performance as we are only concerned about the performance while consuming messages.

File: tests/0089-max_poll_interval.c
Comment (pranavrth) [inline tests/0089-max_poll_interval.c:429]: Can we also write some test case for q forwarding as we are doing that as well? Maybe after the merge if it is going to take time and may affect the release.

**Review by pranavrth — CHANGES_REQUESTED** [review]

File: tests/0089-max_poll_interval.c
Comment (milindl) [reply tests/0089-max_poll_interval.c]: Done

File: src/rdkafka.h
Comment (milindl) [reply src/rdkafka.h]: Done, I got confused as "lose" can also mean get rid of

File: src/rdkafka_queue.h
Comment (milindl) [reply src/rdkafka_queue.h:1186]: We can't do this for correctness reasons even if the flag is only set once. 

Two reasons for this is. First is that that even if the poll is always guaranteed to be called by the user after setting of the flag is complete, it might still be on different threads, and without a memory barrier (like a lock) or unless flags is marked volatile, there is no guarantee that the two threads see the same value of the variable.

Second is that, while I'm reasonably certain that writes and reads from `int` are atomic in most cases, there isn't any requirement for them to be so.

Comment (pranavrth) [reply src/rdkafka_queue.h:1186]: As discussed, First is not a problem. We need to check second one though.

Comment (pranavrth) [reply src/rdkafka_queue.h:1186]: I was suggesting something like _[Double-checked locking](https://en.wikipedia.org/wiki/Double-checked_locking)_

Comment (milindl) [reply src/rdkafka_queue.h:1186]: Resolving for now, will address later to see what we can trade-off wrt performance/correctness.

File: CHANGELOG.md
Comment (emasab) [inline CHANGELOG.md:11]: Not setting `polled` would also make it leave the consumer group I think.
```suggestion
 * Fixes certain cases where polling would not keep the consumer
   in the group or make it rejoin it (#4256).
```

File: CHANGELOG.md
Comment (emasab) [inline CHANGELOG.md:12]:
```suggestion
```

File: CHANGELOG.md
Comment (emasab) [inline CHANGELOG.md:31]:
```suggestion
   in a queue itself, that, whether polling, resets the timer.
```

**Review by emasab — CHANGES_REQUESTED** [review]

File: tests/0089-max_poll_interval.c
Comment (pranavrth) [inline tests/0089-max_poll_interval.c:444]: I think we are just trying to squeeze multiple scenarios in one test case. We can make it different test specific to the case. For the forward queue test, we just need to check if the flag is getting set.

Comment (milindl) [reply tests/0089-max_poll_interval.c:444]: Resolving, as per discussion. We can't access internals to check the flag

**Review by emasab — APPROVED** [review]
Great Milind, it was difficult to find the correct combination of changes that solves both problems by resetting the timer only when needed and fits with the existing code.

**Comment by milindl** [comment]
Thanks Emanuele. Will merge after #4262 for CI

**Review by pranavrth — APPROVED** [review]
LGTM!

**Pull request merged**
